jupyter-docker/jupyter_localize_extension.py

```diff
@@ -43,12 +47,27 @@ def post(self):
     if not all(map(lambda v: type(v) is unicode, pathdict.values())):
       raise HTTPError(400, "Body must be JSON object of type string/string")
 
-    #complete the request HERE, without waiting for the localize to run
-    self.set_status(200)
-    self.finish()
+    synchronous = self.get_query_argument('sync', False)
 
-    #fire and forget the actual work -- it'll log to a file in the user's homedir
-    tornado.ioloop.IOLoop.current().spawn_callback(self.localize, pathdict)
+    if not synchronous:
+      #complete the request HERE, without waiting for the localize to run
+      self.set_status(200)
+      self.finish()
+
+      #fire and forget the actual work -- it'll log to a file in the user's homedir
+      tornado.ioloop.IOLoop.current().spawn_callback(self.localize, pathdict)
+
+    else:
+      #run localize synchronous to the HTTP request
+      #run_sync() doesn't take arguments, so we must wrap the call in a lambda.
+      success = tornado.ioloop.IOLoop().run_sync(lambda: self.localize(pathdict))
+
+      #complete the request only after localize completes
+      if not success:
+        raise HTTPError(500, "Error occurred during localization. See localization.log for details.")
```

Comment:
@zarsky-broad when passing `?sync=true` the API returns 500 if ANY of the copies fail. Reasonable?

Reply:
Sure, I'll wind up calling it once per notebook in that case.

Reply:
(And thank you!)

Reply:
If the main usage requires synchronous localization, I wonder if we should make that the default behaviour.

Reply:
True - it might make more sense to make the default behavior synchronous.